magento/magento2
# URLs should not contains reserved characters according to RFC 3986

### Description (*)


This PR fixes the [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986.html) which forbids to use reserved characters in URLs such as a comma.

### Related Pull Requests


### Manual testing scenarios (*)

1. Use the URL : `http://magento2.adobe/encoding/with/longer/url/`
2. Encode this URL with the `Magento\Framework\Url\Encoder::encode` method
3. Note that the base64 encoded URL contains a comma : `aHR0cDovL21hZ2VudG8yLmFkb2JlL2VuY29kaW5nL3dpdGgvbG9uZ2VyL3VybC8,`
4. Use the native method [rawurlencode](https://www.php.net/manual/fr/function.rawurlencode.php) from PHP on the result : `aHR0cDovL21hZ2VudG8yLmFkb2JlL2VuY29kaW5nL3dpdGgvbG9uZ2VyL3VybC8%2C` (please notice the %2C character which is the hexadecimal value for a comma, which is **not allowed**.
5. Decode this URL back : `http://magento2.adobe/encoding/with/longer/url/6` (a `6` was added at the end of the URL because of the encoded value)

### Questions or comments


In my understanding of the RFC, an URL can't contain a comma (or whatever reserved characters). However, Magento uses such character in the `encode` method to remove the `=` character from the base64 encoded value.

https://github.com/magento/magento2/blob/c6aeb6aef1eaf849aadeeadedd63ac6cd3dc23fc/lib/internal/Magento/Framework/Url/Encoder.php#L18

This PR replaces the comma value by a tilde as this is an unreserved character allowed for an URL and which will not be transformed in the `rawurlencode` method.

### Contribution checklist (*)


### Resolved issues:
1. [x] resolves magento/magento2#36042: URLs should not contains reserved characters according to RFC 3986
